File: bindings/python/tests/test_clock_bridge.py
"""The host-clock read and the ClockBridge built on it.

The point of `bwa_host_time_ns` is that the offset between your clock and the engine's stops being
an estimate: the sandwich bounds its own error, so these tests can assert against that bound rather
than against a tolerance someone picked.
"""

from __future__ import annotations

import time

import bw_audio as bwa
from conftest import BLK, SR


def test_host_time_ns_needs_no_engine_and_advances():
    """It is a property of the process, so it answers before any Engine exists."""
    a = bwa.host_time_ns()
    b = bwa.host_time_ns()
    assert a > 0
    assert b >= a
    # Over a real sleep it must move by about that much, which rules out a stuck or wrong-scale clock.
    t0 = bwa.host_time_ns()
    time.sleep(0.05)
    elapsed_ms = (bwa.host_time_ns() - t0) * 1e-6
    assert 40.0 < elapsed_ms < 500.0, elapsed_ms


def test_offset_is_stable_across_refreshes_within_the_bound(null_engine):
    """|off_i - off_j| <= err_i + err_j is arithmetic, not a tolerance: each sandwich brackets the
    true offset. A failure here means the two clocks are not the same clock, or the read is wrong."""
    c = null_engine.bridge()
    samples = []
    for _ in range(8):
        err = c.refresh()
        samples.append((c.offset_ns, err))
        time.sleep(0.002)
    for i in range(len(samples)):
        for j in range(i + 1, len(samples)):
            oi, ei = samples[i]
            oj, ej = samples[j]
            # 1 us of slack for the ns rounding in the platform conversion.
            assert abs(oi - oj) <= ei + ej + 1000.0, (samples[i], samples[j])


def test_the_error_bound_is_small_on_this_machine(null_engine):
    """Reported, not just asserted: the bound is the number that says how good the anchor is."""
    c = null_engine.bridge()
    print("two-reads error bound: {0:.2f} us".format(c.error_ns * 1e-3))
    assert c.error_ns < 100e3, c.error_ns


def test_perf_counter_is_the_engine_clock(null_engine):
    """On every platform the engine's clock IS time.perf_counter's, so the offset is near zero.

    Not a contract the bridge depends on (it measures whatever offset there is), but it is the
    reason perf_counter is the default, and a large value here would say the platform changed.
    """
    c = null_engine.bridge()
    assert abs(c.offset_ns) < 1e9, c.offset_ns


def test_dsp_at_and_time_at_round_trip(manual_engine):
    e = manual_engine
    for _ in range(4):
        e.render_block()
    c = e.bridge()
    for lead in (0.0, 0.01, 0.5, 5.0):
        t = time.perf_counter() + lead
        assert abs(c.time_at(c.dsp_at(t)) - t) < 1e-3, lead


def test_dsp_at_of_the_stamps_own_host_time_is_the_stamps_sample(manual_engine):
    """The manual sink's stamps are synthesized from the sample position, so this is exact.

    Feed the bridge the caller-clock time that maps to the pair's own host time and it must hand
    back the pair's own sample, with no drift term and no rounding past a sample.
    """
    e = manual_engine
    for _ in range(4):
        e.render_block()
    pair = e.clock
    assert pair is not None
    cs, _ct = pair
    c = e.bridge()
    assert abs(c.dsp_at(c.time_at(cs)) - cs) <= 1


def test_a_caller_clock_with_a_huge_epoch_offset_works(manual_engine):
    """The bridge measures the offset, so an arbitrary epoch costs nothing."""
    e = manual_engine
    for _ in range(4):
        e.render_block()
    shift = 1.0e6                       # a clock whose zero is 11 days from perf_counter's
    plain = e.bridge()
    shifted = e.bridge(clock=lambda: time.perf_counter() + shift)
    t = time.perf_counter() + 0.1
    assert abs(shifted.offset_ns - (plain.offset_ns - shift * 1e9)) < 1e6
    # Same instant described in two clocks must land on the same sample, to within a block.
    assert abs(shifted.dsp_at(t + shift) - plain.dsp_at(t)) < BLK


def test_dsp_at_answers_before_the_first_stamped_block(null_engine):
    """No pair yet is a fallback, not a failure: pair the block counter with the caller's clock."""
    c = null_engine.bridge()
    assert null_engine.clock is None
    assert c.dsp_at(time.perf_counter()) >= 0
    assert c.dsp_at(time.perf_counter() - 1e6) == 0      # a past time clamps rather than going negative


def test_dsp_at_now_tracks_the_engine_clock():
    """On a host-paced sink the stamps are real host times, so "now" on the caller's clock must be
    the engine's own now. This is the assertion the whole call exists for."""
    e = bwa.Engine(profile=bwa.Profile.CAVE, sample_rate=SR, block_size=BLK, sink=bwa.SinkType.NULL)
    try:
        e.start()
        deadline = time.perf_counter() + 2.0
        while e.clock is None and time.perf_counter() < deadline:
            time.sleep(0.01)
        assert e.clock is not None, "the null sink never stamped a block"
        c = e.bridge()
        err = c.dsp_at(time.perf_counter()) - e.dsp_time_frames
        print("dsp_at(now) - dsp_time_frames: {0} frames".format(err))
        # dsp_time_frames is the LAST block's first sample, so "now" is legitimately up to a block
        # and a bit ahead of it. Four blocks covers a preempted interpreter; an epoch error would
        # be millions.
        assert -BLK < err < 4 * BLK, err
    finally:
        e.close()


def test_play_at_takes_off_the_output_latency(manual_engine, tmp_path):
    from conftest import tone, write_wav

    e = manual_engine
    snd = e.load_sound(write_wav(tmp_path / "click.wav", tone(SR // 10)))
    src = e.create_source()
    src.set_pos(0.0, 1.5, 2.0)
    e.commit()
    for _ in range(4):
        e.render_block()

    c = e.bridge()
    t = c.time_at(e.dsp_time_frames + 10 * BLK)
    want = c.dsp_at(t)
    start = c.play_at(src, snd, t)
    assert start == max(0, want - e.output_latency_frames)
    assert start > e.dsp_time_frames                      # still in the future, so it is schedulable


def test_drift_ppm_is_none_or_a_number(manual_engine):
    """The fit needs about a second of stamps. Before that the property says so rather than lying."""
    ppm = manual_engine.bridge().drift_ppm
    assert ppm is None or isinstance(ppm, float)


def test_refresh_returns_the_bound_it_set(null_engine):
    c = null_engine.bridge()
    err = c.refresh(rounds=3)
    assert err == c.error_ns
    assert err > 0.0


def test_bridge_repr_is_ascii(null_engine):
    r = repr(null_engine.bridge())
    assert r.isascii()
    assert "ClockBridge" in r

File: bindings/matlab/tests/test_clock_bridge.m
function test_clock_bridge()
%TEST_CLOCK_BRIDGE  bwa.ClockBridge: the sandwich, the round trip, and the exact case.
%
%   The manual sink is what makes this testable. Its (sample, host time) stamp is SYNTHESIZED
%   from the sample position, so the pair is exact arithmetic rather than a measurement, and
%   dsp_at of a stamp's own host time has to come back as that stamp's sample to the sample.
%   On hardware the same identity holds only to the stamp's own jitter.

C = bwa.Const;
SR = 48000;
BLK = 256;

e = bwa.Engine('profile', C.PROFILE_CAVE, 'sink', C.SINK_MANUAL, ...
               'sample_rate', SR, 'block_size', BLK);
cleanup = onCleanup(@() e.close()); %#ok<NASGU>

% host_time_ns needs no engine and must advance.
t1 = bwa_mex('host_time_ns');
pause(0.02);
t2 = bwa_mex('host_time_ns');
tcheck('host_time_ns returns a uint64', isa(t1, 'uint64'));
tcheck('host_time_ns advances', t2 > t1, sprintf('%d then %d', t1, t2));
tcheck('host_time_ns advances by about the wait', ...
       double(t2 - t1) * 1e-9 > 0.005 && double(t2 - t1) * 1e-9 < 2.0, ...
       sprintf('%.1f ms', double(t2 - t1) * 1e-6));
tcheck('the engine method agrees with the raw call', e.host_time_ns() >= t2);

b = e.bridge();
tcheck('the bridge builds before start', isa(b, 'bwa.ClockBridge'));

% The sandwich is two reads of the caller's clock around one of the engine's, so its WIDTH is
% the error bound on the offset. Report it: a number is worth more here than a threshold.
% Measure the STEADY STATE over a few refreshes rather than one reading. A cold first call into
% a MEX pays its load cost inside the sandwich and reads a millisecond wide, which says nothing
% about what a trial loop will see. The constructor already discards one for that reason.
widths = zeros(1, 5);
for k = 1:5
    b.refresh();
    widths(k) = b.sandwich_seconds;
end
fprintf('   (the offset sandwich measured %.2f us, best of 5; worst %.2f us)\n', ...
        min(widths) * 1e6, max(widths) * 1e6);
tcheck('the sandwich is not negative', all(widths >= 0));
tcheck('the sandwich is well under a millisecond', min(widths) < 2e-4, ...
       sprintf('%.2f us', min(widths) * 1e6));

% Before a stamped block the bridge still answers, block-granular, rather than refusing.
tcheck('the bridge answers before the first stamped block', b.dsp_at(b.now()) >= 0);
tcheck('a past time clamps at zero', b.dsp_at(b.now() - 1e6) == 0);

e.start();
for k = 1:8
    e.render_block();
end
b.refresh();
tcheck('the bridge sees a stamped pair once blocks have rendered', b.valid);

% THE identity: dsp_at of the stamp's own host time is the stamp's own sample. The bridge stores
% the pair on the engine's clock, so feeding it back through the caller-clock conversion has to
% return where it started. On the manual sink that is exact.
own_t = b.host - b.offset;             % the stamp's host time, expressed on the caller's clock
tcheck('dsp_at of a stamp''s own host time is that stamp''s sample', ...
       b.dsp_at(own_t) == b.sample, sprintf('%d against %d', b.dsp_at(own_t), b.sample));

% And the inverse.
tcheck('time_at inverts dsp_at at the stamp', abs(b.time_at(b.sample) - own_t) < 1e-9, ...
       sprintf('%.3e s', abs(b.time_at(b.sample) - own_t)));
probe = own_t + 0.25;
tcheck('time_at inverts dsp_at away from the stamp', ...
       abs(b.time_at(b.dsp_at(probe)) - probe) < 1e-4, ...
       sprintf('%.3e s', abs(b.time_at(b.dsp_at(probe)) - probe)));

% A quarter second ahead is a quarter second of samples ahead, at the engine's own rate.
ahead = b.dsp_at(own_t + 0.25) - b.sample;
tcheck('a quarter second ahead is a quarter second of samples', ...
       abs(double(ahead) - 0.25 * SR) <= 2, sprintf('%d against %d', ahead, 0.25 * SR));

% The manual sink's clock model fits ppm = 0 exactly, which is true of that fiction and of no
% hardware. If it is there, the bridge should be using its rate.
m = e.clockModel();
if ~isempty(m)
    tcheck('the manual sink fits zero drift', abs(m.ppm) < 1e-6, sprintf('%g ppm', m.ppm));
    tcheck('the bridge adopted the fitted rate', abs(b.rate_hz - m.rate_hz) < 1e-9);
else
    tcheck('the bridge falls back to the nominal rate with no fit', b.rate_hz == 0);
end

% play_at subtracts the device latency and nothing else. With no physical output that is 0, so
% the scheduled sample IS dsp_at, which is the arithmetic worth pinning.
tmp = tempname();
mkdir(tmp);
cleandir = onCleanup(@() rmdirQuiet(tmp)); %#ok<NASGU>
wav = fullfile(tmp, 'burst.wav');
audiowrite(wav, 0.9 * ones(BLK, 1), SR);
snd = e.loadSound(wav);
src = e.createSource();
src.setPos(0, 1.5, 2);

target = b.time_at(double(e.dsp_time_frames) + 10 * BLK);
start = b.play_at(src, snd, target);
tcheck('play_at returns the sample it scheduled', start > 0, num2str(start));
tcheck('play_at subtracts the output latency', ...
       start == b.dsp_at(target) - double(e.output_latency_frames), num2str(start));

% And it really lands there: the blocks before it are silent and the one at it is not.
peaks = zeros(1, 14);
for k = 1:14
    out = e.render_block();
    peaks(k) = max(abs(out(:)));
end
firstLoud = find(peaks > 1e-6, 1);
tcheck('the bridge-scheduled sound started', ~isempty(firstLoud));
if ~isempty(firstLoud)
    tcheck('nothing sounded before the scheduled block', firstLoud >= 9, num2str(firstLoud));
    tcheck('it sounded within a block of the schedule', firstLoud <= 12, num2str(firstLoud));
end

% A custom clock function is honored, which is what a Psychtoolbox rig will pass.
b2 = e.bridge(@() 1000.0);
tcheck('a custom clock is used', b2.now() == 1000.0);
% The offset is measured against host_time_ns at refresh, not against the stamped pair, so a
% clock frozen at 1000 puts the whole of the engine's host clock into it.
h_now = double(bwa_mex('host_time_ns')) * 1e-9;
tcheck('a custom clock gives its own offset', abs(b2.offset - (h_now - 1000.0)) < 0.5, ...
       sprintf('%.6f against %.6f', b2.offset, h_now - 1000.0));
tcheck('a clock that cannot move measures a zero-width sandwich', b2.sandwich_seconds == 0);
end

function rmdirQuiet(d)
try
    rmdir(d, 's');
catch
end
end
